crates/bevy_reflect/src/reflect.rs

```diff
@@ -86,3 +90,49 @@ impl dyn Reflect {
         self.any_mut().downcast_mut::<T>()
     }
 }
+
+unsafe impl Reflect for Box<dyn Reflect> {
+    fn type_name(&self) -> &str {
+        self.deref().type_name()
+    }
+
+    fn any(&self) -> &dyn Any {
+        self.deref().any()
```

Comment:
Is this safe/correct? The unsafe contract for `impl Reflect` is that `any` and `any_mut` return the `self` value passed in. I could see this being correct but I could also see an edge case like boxing the `Box<dyn Reflect>` and using `downcast` on that being UB.

Reply:
Some solutions I can think of:
- Just return `self` instead of `self.deref().any()`
- Add `any_box` to `Reflect` which takes `self` and returns `Box<dyn Any>`, and then remove the unsafe in `downcast`, which should also make Reflect entirely safe to implement
- Move downcast to be a `Reflect` method instead of an impl on `dyn Reflect`?